File: shared/AgentAreaTests.cs
// Copyright (c) 2026 DOTS
// Cross-platform tests for Chat/Coding persistence and Chat context boundaries.

using System.Text.Json;
using DotsHarnessCore;
using PluginRuntime;
using Xunit;

namespace DotsHarness.Tests;

public sealed class AgentAreaTests : IDisposable
{
    private readonly string _root = Path.Combine(
        Path.GetTempPath(),
        "DotsHarnessAreaTests-" + Guid.NewGuid().ToString("N"));
    private readonly SupportPaths _paths;

    public AgentAreaTests()
    {
        _paths = new SupportPaths(
            _root,
            Path.Combine(_root, "plugins"),
            Path.Combine(_root, "presets"),
            Path.Combine(_root, "settings.json"),
            Path.Combine(_root, "host.patch.yml"),
            Path.Combine(_root, "trust.json"),
            Path.Combine(_root, "models"),
            Path.Combine(_root, "runtime"));
        _paths.Ensure();
    }

    [Fact]
    public void LegacySessionsSplitByWorkspaceAndKeepTheLegacyFile()
    {
        var workspace = Path.Combine(_root, "coding-project");
        Directory.CreateDirectory(workspace);
        var legacyPath = Path.Combine(_paths.Root, "sessions.json");
        var legacy = new[]
        {
            new Conversation { Id = "coding", Cwd = workspace },
            new Conversation { Id = "chat", Cwd = "", Area = AgentArea.Coding },
        };
        File.WriteAllText(legacyPath, JsonSerializer.Serialize(legacy));

        SessionAreaMigration.Run(_paths);

        var coding = Read(Path.Combine(_paths.Root, AgentArea.Coding.SessionFileName()));
        var chat = Read(Path.Combine(_paths.Root, AgentArea.Chat.SessionFileName()));
        Assert.Equal("coding", Assert.Single(coding).Id);
        Assert.Equal(AgentArea.Coding, coding[0].Area);
        Assert.Equal("chat", Assert.Single(chat).Id);
        Assert.Equal(AgentArea.Chat, chat[0].Area);
        Assert.Null(chat[0].Cwd);
        Assert.True(File.Exists(legacyPath));
        Assert.True(File.Exists(Path.Combine(_paths.Root, "area-session-migration-v1.json")));
    }

    [Fact]
    public void PathParserRecognizesPosixWindowsDriveAndUncCandidates()
    {
        var extracted = ChatPathParser.ExtractPaths(
            "POSIX /tmp/project/main.cs, drive C:\\repo\\main.cs and UNC \\\\server\\share\\main.cs.");

        Assert.Contains("/tmp/project/main.cs", extracted);
        Assert.Contains("C:\\repo\\main.cs", extracted);
        Assert.Contains(@"\\server\share\main.cs", extracted);
    }

    [Fact]
    public void LedgerKeepsMultipleRootsAndCacheNamespacesDoNotContainRawPaths()
    {
        var first = Path.Combine(_root, "project-a");
        var second = Path.Combine(_root, "project-b");
        Directory.CreateDirectory(first);
        Directory.CreateDirectory(second);
        var ledger = new ChatContextLedger();
        var attached = ledger.Attach(
            [
                ChatContextRoot.Create(first, ChatContextRootKind.Directory, "message-a"),
                ChatContextRoot.Create(second, ChatContextRootKind.Directory, "message-b"),
            ],
            "message-a");

        Assert.Equal(2, attached.Count);
        Assert.Equal(2, ledger.Roots.Count);
        Assert.Contains("message-a", ledger.Roots[0].SourceMessageIds);
        Assert.Contains("message-b", ledger.Roots[1].SourceMessageIds);

        var chatKey = AgentCacheNamespace.Create(
            AgentArea.Chat, "conversation", "chat-project", null, "model", false, ledger.Revision,
            ledger.Roots.Select(root => root.Id));
        var codingKey = AgentCacheNamespace.Create(
            AgentArea.Coding, "conversation", null, "coding-project", "model", false, ledger.Revision);

        Assert.NotEqual(chatKey, codingKey);
        Assert.DoesNotContain(first, chatKey, StringComparison.Ordinal);
        Assert.DoesNotContain(second, chatKey, StringComparison.Ordinal);
    }

    [Fact]
    public void ProviderAccountsUseOneSharedStoreAcrossAreas()
    {
        var model = new AppModel(_paths);

        Assert.Same(model.ChatRouter.Store, model.CodingRouter.Store);
        Assert.NotSame(model.ChatBridge.Skills, model.CodingBridge.Skills);
        Assert.Same(model.ActiveSkills, model.ChatBridge.Skills);
        model.SetActiveArea(AgentArea.Coding);
        Assert.Same(model.ActiveSkills, model.CodingBridge.Skills);
    }

    [Fact]
    public void LastActiveAreaSurvivesAppModelRecreation()
    {
        var firstLaunch = new AppModel(_paths);
        firstLaunch.SetActiveArea(AgentArea.Coding);
        Assert.Equal(AgentArea.Coding, firstLaunch.ActiveArea);

        var codingLaunch = new AppModel(_paths);
        Assert.Equal(AgentArea.Coding, codingLaunch.ActiveArea);

        codingLaunch.SetActiveArea(AgentArea.Chat);
        Assert.Equal(AgentArea.Chat, codingLaunch.ActiveArea);

        var chatLaunch = new AppModel(_paths);
        Assert.Equal(AgentArea.Chat, chatLaunch.ActiveArea);
    }

    [Fact]
    public async Task ChatToolsRequireAValidRootAndStayInsideIt()
    {
        var workspace = Path.Combine(_root, "chat-root");
        Directory.CreateDirectory(workspace);
        var inside = Path.Combine(workspace, "inside.txt");
        var outside = Path.Combine(_root, "outside.txt");
        File.WriteAllText(inside, "inside");
        File.WriteAllText(outside, "outside");
        var root = ChatContextRoot.Create(workspace, ChatContextRootKind.Directory, "message");

        var outsideCall = new NativeToolCall(
            "read-outside",
            "read_file",
            JsonSerializer.Serialize(new { contextRootID = root.Id, path = outside }));
        var outsideResult = await NativeWorkspaceTools.ExecuteAsync(outsideCall, [root]);
        Assert.Contains("outside the selected context root", outsideResult, StringComparison.OrdinalIgnoreCase);

        var invalidRootCall = new NativeToolCall(
            "read-invalid",
            "read_file",
            JsonSerializer.Serialize(new { contextRootID = "unknown", path = "inside.txt" }));
        var invalidRootResult = await NativeWorkspaceTools.ExecuteAsync(invalidRootCall, [root]);
        Assert.Contains("valid contextRootID", invalidRootResult, StringComparison.OrdinalIgnoreCase);

        var escapingCommand = new NativeToolCall(
            "command-escape",
            "run_command",
            JsonSerializer.Serialize(new { contextRootID = root.Id, command = "cd .. && pwd" }));
        var commandResult = await NativeWorkspaceTools.ExecuteAsync(escapingCommand, [root]);
        Assert.Contains("leave the selected context root", commandResult, StringComparison.OrdinalIgnoreCase);
    }

    [Fact]
    public async Task FileContextRootCanOnlyAddressThatFile()
    {
        var workspace = Path.Combine(_root, "file-root");
        Directory.CreateDirectory(workspace);
        var file = Path.Combine(workspace, "answer.txt");
        var other = Path.Combine(workspace, "other.txt");
        File.WriteAllText(file, "answer");
        File.WriteAllText(other, "other");
        var root = ChatContextRoot.Create(file, ChatContextRootKind.File, "message");

        var read = new NativeToolCall(
            "read-file",
            "read_file",
            JsonSerializer.Serialize(new { contextRootID = root.Id }));
        Assert.Equal("answer", await NativeWorkspaceTools.ExecuteAsync(read, [root]));

        var list = new NativeToolCall(
            "list-file",
            "list_files",
            JsonSerializer.Serialize(new { contextRootID = root.Id, path = "answer.txt" }));
        var listResult = await NativeWorkspaceTools.ExecuteAsync(list, [root]);
        Assert.Contains("requires a directory context root", listResult, StringComparison.OrdinalIgnoreCase);

        var otherRead = new NativeToolCall(
            "read-other",
            "read_file",
            JsonSerializer.Serialize(new { contextRootID = root.Id, path = "other.txt" }));
        var otherResult = await NativeWorkspaceTools.ExecuteAsync(otherRead, [root]);
        Assert.Contains("outside the selected context root", otherResult, StringComparison.OrdinalIgnoreCase);
    }

    private static List<Conversation> Read(string path) =>
        JsonSerializer.Deserialize<List<Conversation>>(File.ReadAllText(path)) ?? [];

    public void Dispose()
    {
        try
        {
            if (Directory.Exists(_root)) Directory.Delete(_root, recursive: true);
        }
        catch { }
    }
}

File: macos/DotsHarness/Tests/HarnessPluginKitTests/AgentAreaTests.swift
import Foundation
import XCTest
import PluginRuntime
@testable import DotsHarnessCore

final class AgentAreaTests: XCTestCase {
    func testLedgerRetainsMultipleRootsAndStructuredMemory() {
        var ledger = ChatContextLedger()
        let first = ChatContextRoot(path: "/tmp/project-a", kind: .directory)
        let second = ChatContextRoot(path: "/tmp/project-b", kind: .directory)

        XCTAssertEqual(ledger.attach(roots: [first], to: "message-a").count, 1)
        XCTAssertEqual(ledger.attach(roots: [second], to: "message-b").count, 1)
        ledger.addProjectNote("Keep the two roots independent")
        ledger.addChangeSummary("Updated project-a")

        XCTAssertEqual(ledger.roots.map(\.path), [first.path, second.path])
        XCTAssertEqual(ledger.roots[0].sourceMessageIDs, ["message-a"])
        XCTAssertEqual(ledger.projectNotes, ["Keep the two roots independent"])
        XCTAssertEqual(ledger.changeSummaries, ["Updated project-a"])
        XCTAssertGreaterThanOrEqual(ledger.revision, 4)
    }

    func testPathParserSupportsQuotedCrossPlatformCandidatesAndKnownRelativeRoots() throws {
        let paths = temporaryPaths()
        let root = paths.root.appendingPathComponent("relative-root", isDirectory: true)
        let source = root.appendingPathComponent("Sources/App.swift")
        try FileManager.default.createDirectory(at: source.deletingLastPathComponent(), withIntermediateDirectories: true)
        try Data("print(\"ok\")".utf8).write(to: source)
        defer { try? FileManager.default.removeItem(at: root) }

        let text = "Inspect \"\(root.path)\", `\(source.path)` and C:\\repo\\App.swift from \\\\server\\share\\repo."
        let candidates = ChatPathParser.extractPaths(from: text)
        XCTAssertTrue(candidates.contains(root.path))
        XCTAssertTrue(candidates.contains(source.path))
        XCTAssertTrue(candidates.contains("C:\\repo\\App.swift"))
        XCTAssertTrue(candidates.contains("\\\\server\\share\\repo"))

        let relative = ChatPathParser.roots(
            from: "Please inspect Sources/App.swift",
            knownRoots: [ChatContextRoot(path: root.path, kind: .directory)],
            sourceMessageID: "relative-message"
        )
        XCTAssertEqual(relative.map(\.path), [source.resolvingSymlinksInPath().path])
        XCTAssertEqual(relative.first?.sourceMessageIDs, ["relative-message"])
    }

    func testLegacySessionsSplitByWorkspaceWithoutDeletingLegacyFile() throws {
        let paths = temporaryPaths()
        let codingRoot = paths.root.appendingPathComponent("coding-root", isDirectory: true)
        try FileManager.default.createDirectory(at: codingRoot, withIntermediateDirectories: true)
        defer { try? FileManager.default.removeItem(at: codingRoot) }

        let legacy: [[String: Any]] = [
            ["id": "coding-session", "title": "Code", "messages": [], "cwd": codingRoot.path],
            ["id": "chat-session", "title": "Chat", "messages": []],
        ]
        let legacyURL = paths.root.appendingPathComponent("sessions.json")
        try JSONSerialization.data(withJSONObject: legacy).write(to: legacyURL)

        SessionAreaMigration.run(paths: paths)

        let coding = try decodeConversations(at: paths.root.appendingPathComponent(AgentArea.coding.sessionFileName))
        let chat = try decodeConversations(at: paths.root.appendingPathComponent(AgentArea.chat.sessionFileName))
        XCTAssertEqual(coding.map(\.id), ["coding-session"])
        XCTAssertEqual(coding.first?.area, .coding)
        XCTAssertEqual(coding.first?.cwd, codingRoot.path)
        XCTAssertEqual(chat.map(\.id), ["chat-session"])
        XCTAssertEqual(chat.first?.area, .chat)
        XCTAssertNil(chat.first?.cwd)
        XCTAssertTrue(FileManager.default.fileExists(atPath: legacyURL.path))
    }

    func testCacheNamespaceSeparatesAreasAndContextRevisionsWithoutRawPaths() {
        let chat = AgentCacheNamespace.key(
            area: .chat,
            conversationID: "conversation",
            chatProjectID: "chat-project",
            codingProjectID: nil,
            model: "model",
            planMode: false,
            contextRevision: 1,
            rootIDs: ["context-a"]
        )
        let coding = AgentCacheNamespace.key(
            area: .coding,
            conversationID: "conversation",
            chatProjectID: nil,
            codingProjectID: "coding-project",
            model: "model",
            planMode: false,
            contextRevision: 1
        )
        let nextRevision = AgentCacheNamespace.key(
            area: .chat,
            conversationID: "conversation",
            chatProjectID: "chat-project",
            codingProjectID: nil,
            model: "model",
            planMode: false,
            contextRevision: 2,
            rootIDs: ["context-a"]
        )

        XCTAssertNotEqual(chat, coding)
        XCTAssertNotEqual(chat, nextRevision)
        XCTAssertFalse(chat.contains("/tmp/"))
    }

    @MainActor
    func testAppModelKeepsChatAndCodingConversationsAndDraftsSeparate() throws {
        let paths = temporaryPaths()
        let codingRoot = paths.root.appendingPathComponent("coding-root", isDirectory: true)
        try FileManager.default.createDirectory(at: codingRoot, withIntermediateDirectories: true)
        defer { try? FileManager.default.removeItem(at: codingRoot) }

        let model = AppModel(paths: paths)
        XCTAssertEqual(model.activeArea, .chat)
        XCTAssertTrue(model.chatRouter.store === model.codingRouter.store)
        XCTAssertFalse(model.chatBridge.skills === model.codingBridge.skills)
        XCTAssertNotNil(model.chatBridge.skillSuggestions)
        XCTAssertNotNil(model.codingBridge.skillSuggestions)
        XCTAssertTrue(model.activeSkills === model.chatBridge.skills)

        model.setWorkspace(codingRoot.path)
        XCTAssertEqual(model.activeArea, .coding)
        XCTAssertTrue(model.activeSkills === model.codingBridge.skills)
        let codingID = try XCTUnwrap(model.selectedConversationID)
        model.draft = "coding draft"

        model.setActiveArea(.chat)
        XCTAssertEqual(model.activeArea, .chat)
        model.draft = "chat draft"
        model.newChatConversation()
        let chatID = try XCTUnwrap(model.selectedConversationID)

        XCTAssertNotEqual(chatID, codingID)
        XCTAssertTrue(model.codingBridge.conversations.contains { $0.id == codingID })
        XCTAssertTrue(model.chatBridge.conversations.contains { $0.id == chatID })
        XCTAssertTrue(model.chatBridge.conversations.allSatisfy { $0.area == .chat && $0.cwd == nil })

        model.setActiveArea(.coding)
        XCTAssertEqual(model.draft, "coding draft")
        XCTAssertEqual(model.selectedConversationID, codingID)
        model.setActiveArea(.chat)
        XCTAssertEqual(model.draft, "chat draft")
        XCTAssertEqual(model.selectedConversationID, chatID)
    }

    @MainActor
    func testAppModelRestoresLastActiveAreaAfterRecreation() throws {
        let paths = temporaryPaths()

        let firstLaunch = AppModel(paths: paths)
        firstLaunch.setActiveArea(.coding)
        XCTAssertEqual(firstLaunch.activeArea, .coding)

        let codingLaunch = AppModel(paths: paths)
        XCTAssertEqual(codingLaunch.activeArea, .coding)

        codingLaunch.setActiveArea(.chat)
        XCTAssertEqual(codingLaunch.activeArea, .chat)

        let chatLaunch = AppModel(paths: paths)
        XCTAssertEqual(chatLaunch.activeArea, .chat)
    }

    private func decodeConversations(at url: URL) throws -> [Conversation] {
        try JSONDecoder().decode([Conversation].self, from: Data(contentsOf: url))
    }

    private func temporaryPaths() -> SupportPaths {
        let root = FileManager.default.temporaryDirectory
            .appendingPathComponent("DotsHarnessAgentAreaTests-\(UUID().uuidString)", isDirectory: true)
        let paths = SupportPaths(
            root: root,
            plugins: root.appendingPathComponent("plugins", isDirectory: true),
            presets: root.appendingPathComponent("presets", isDirectory: true),
            settings: root.appendingPathComponent("settings.json"),
            hostPatch: root.appendingPathComponent("host.patch.yml"),
            trust: root.appendingPathComponent("trust.json"),
            models: root.appendingPathComponent("models", isDirectory: true),
            runtime: root.appendingPathComponent("runtime", isDirectory: true)
        )
        paths.ensure()
        return paths
    }
}
